### PR TITLE
TextViewer: add dialog to set font size and justify text

### PR DESCRIPTION
**Pull request by hius07**
- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11210)

### CODE REVIEW

File: frontend/ui/widget/textviewer.lua [hunk 1]
```diff
@@ -531,6 +540,90 @@ function TextViewer:handleTextSelection(text, hold_duration, start_idx, end_idx,
...
+                UIManager:show(widget)
+            end,
+            hold_callback = function()
+                TextViewer.saveSetting("textviewer_font_size", self.text_font_size)
+            end,
+        }},
+        {{
+            text_func = function()
+                return _("Monospace font") .. (self.monospace_font and "  \u{2713}" or "  \u{2003}")
+            end,
```
Comment (poire-z) [inline frontend/ui/widget/textviewer.lua:590]: Dunno if you witnessed some odd stuff while testing, but sure you don't need the ugly trick I had to use with this checkbox:
https://github.com/koreader/koreader/blob/13a521c39849459cfbbd69e01f17c0c601d7dfbd/frontend/ui/widget/bookmapwidget.lua#L1204-L1229

Or is it because all such checkboxes calls :reinit(), which will rebuilt this popup menu ?
(If yes, I guess everything is rebuild and redrawn, and that the menu popup is closed as soon as you toggle/change things ? Which is a bit different than the behaviour in PageBrowser/Bookmap where we just update the content, and the popup menu stays open. But ok if it's too complicated to get this same behaviour?)

Comment (hius07) [reply frontend/ui/widget/textviewer.lua:590]: Yes, the menu is closed on tap.
If we tap Font size, we see the SpinWidget (that remains open on apply) and we need to see more text to choose the size.
If we tap Monospace or Justify, we see the immediate effect on text, and what is our most frequent next action?
If it is reading, that's okay to close the menu.
If it is switching the setting back, it's better to keep it open.
If it is long-pressing to set the setting as default, it's better to keep it open.

File: frontend/ui/widget/textviewer.lua [hunk 2]
```diff
@@ -57,8 +58,10 @@ local TextViewer = InputContainer:extend{
     title_multilines = nil, -- see TitleBar for details
     title_shrink_font_to_fit = nil, -- see TitleBar for details
 
+    show_menu = true, -- titlebar left icon
     text_font_face = nil, -- default "x_smallinfofont"
     text_font_size = nil,
+    monospace_font = nil, -- "infont"
```
[the remark this answers is not in the document]
Comment (hius07) [reply frontend/ui/widget/textviewer.lua:64]: https://github.com/koreader/koreader/blob/13a521c39849459cfbbd69e01f17c0c601d7dfbd/frontend/ui/widget/infomessage.lua#L50
I'll update the comment.

**Comment by hius07** [comment]
Menu remains open on tapping Monospace and Justify.

**Comment by poire-z** [comment]
I see you moved the checkmark logic in Button - and that maybe we won't have to use my sizing trick in Pagebrowser and Bookmap.
But don't you also need to size it as if the checkmark was enabled (even if it is not initially) in Button?

**Comment by hius07** [comment]
Empty space is reserved with "wide em space" as you did.

**Comment by hius07** [comment]
I didn't tested long texts with checkmark, I suppose the checkmark will be truncated.
Maybe someday we will rewrite the button to have the checkmark at the left, like the touchmenu items.

**Comment by poire-z** [comment]
> Empty space is reserved with "wide em space" as you did.

Oh, right, didn't catch my eyes :) `and "  \u{2713}" or "  \u{2003}")`.
Definitely needs some comment to tell which char these are and why we do that.

> I didn't test long texts with checkmark, I suppose the checkmark will be truncated.

I guess no need to test for text longer than the screen.
Your "Monospace font" is the longest of the 3 items, so you're already in the situation to test the case I needed my trick for.

> Maybe someday we will rewrite the button to have the checkmark at the left, like the touchmenu items.

Not sure I would prefer that: in such top left popup menu, where items are left aligned, there might be 3 kinds of items:
- plain test (your `Font size: 20`)
- text with checkmarks
- `text with | - | +` on the right

I think it looks better if all the text are aligned on the left. If you put the checkmark on the left for the 2nd kind, it would feel quite unaligned/ragged/odd.

**Comment by hius07** [comment]
> If you put the checkmark on the left for the 2nd kind, it would feel quite unaligned/ragged/odd.

We would need a left margin for all the items (again, like in the touchmenu).
(I'm not very enthusiastic to do it right now)

**Comment by poire-z** [comment]
> We would need a left margin for all the items (again, like in the touchmenu).

Yes, some indentation even when it is not needed, that may look odd, and could make that popup menu a bit wider, hiding a bit more of the content below.

It works on our TouchMenu as it takes the full width, and there's blank on the right of most items, so it may balance the blank at start. In this popup anchored at top left, it may look odd, or we may need to add the same right marging for visual niceness, making it even wider.

> (I'm not very enthusiastic to do it right now)

Good ! :)

**Comment by hius07** [comment]
Do I fix your widgets here?

**Comment by poire-z** [comment]
> Do I fix your widgets here?

I guess I would have liked it to result in an individual commit in master once merged (which you can't do the way you work via github web), but I guess it will be mostly code removal, so it might be ok if it all happens in one commit. So, as you feel.

Tested on my Kobo, and the checkmark and sizing look fine, and the fact that it is kept open is really practical :)
(Wondering if the font size could be a `| - | + |` instead of a SpinWidget - but I guess then it's harder to find out something to allow to set it as default.)

But i noticed this: if you set font size (from 20 to 16), it is displayed in 16. Then, if you toggle `Justify`, it stays in 16. But if you toggle `Monospace font` it is unexpectedly brought back to 20.

**Comment by hius07** [comment]
Here is the question: if a caller sets font_face but does not set font_size, should TextViewer use font_size from G_settings, or use the default size of font_face?

**Comment by poire-z** [comment]
> Here is the question: if a caller sets font_face but does not set font_size, should TextViewer use font_size from G_settings, or use the default size of font_face?

Caller = some code specifying a specific font face ? Do we have some ?
Or the user toggling "Monospace font" ?
I guess it should use the G_reader_setting set by the user.

If it's about HTML and CSS viewer, which forces it to monospace, I dunno :) I would like these to be kept small so I see more of the HTML/CSS - and them to not be affected by the probably larger font size I have set for Book description and other non-code content.

**Comment by poire-z** [comment]
Ideally, all TextViewer instances could have a "category" attribute, ie category="book_description", "highlight", "html_viewer", "css_viewer" - and the default would be saved/associated with that category - or even no need for a default: just save and reuse the last font size set.
And I guess this should also apply to justify/monospace...

**Comment by hius07** [comment]
Setting `font_face ` in a caller overrides both `font_face` and `font_size` from TextViewer G_settings (to avoid changes to ViewHtml and ReaderTypography).

**Comment by poire-z** [comment]
I guess that with my "Ideally..." above, the font_face we set in the code for specific cases would be some initial suggested/best options for that context - but a user then tweaking them (if things are saved by category) could tweak them and get them reused next time it goes to that TextViewer.

We could save and reuse:
```lua
["textviewer_settings"] = {
  "css_viewer" = { "font_size" = 16, monospace = true, justify = false ,
  "highlight" = { "font_size" = 22, monospace = false, justify = true },
   ...
}
```

Not asking you to do that in this PR, just sharing thoughts for discussion and some possible next step.

**Comment by poire-z** [comment]
We may have 2 different "looks" with out defaults (that is: for code, and for content related text), which is probably fine. (I would consider the somehow hidden one in ReaderTypography as code).

But a user may like to tweak the 10 ones differently, for his own reasons - ie. highlight/notes usually have some small text, so a bigger font is ok. Book description may be longer, so a smaller font might be better. Translator may depend on how the user usually uses it: for single word lookup as a dict, and a larger font might work - or to translate full paragraphs and get many translation variants, and a smaller font might be better :)

So, it depends on how we want to go:
If only 2 categories (code vs. book-related-text), we could just have a flag in the TextViewer instances, and the user would need to tweak each of the 2 kind only once - and get it everywhere.
If we allow more categories (say 7), it brings more flexibility, but the user may have to tweak it the same 7 times.

I guess I would be fine with any of these choices, but my personnal "tweaker" preference would be for flexibility.
(Also, I have other TextViewer instances in my own patch, like for viewing logs, where I would like to use a monospace font smaller than the one I'll get for HTML/CSS viewing - so it would be nice if you could still allow that :))

**Comment by hius07** [comment]
It's worth thinking over, for a future PR.

Categories and their default TextViewer settings (draft)

**code** (16, m+, j-)
(01) ViewHtml: code
(02) ReaderTypography: "Language tags (and hyphenation dictionaries) used since start up"
(03) PatchManagement: code

**dictionary** (20, m-, j-)
(04) Translator: translation

**bookinfo** (20, m-, j+)
(05) FilemanagerBookInfo: book description
(06) OPDSBrowser: book information

**bookmark** (20, m-, j+)
(07) ReaderBookmark: bookmark details
(08) ReaderHighlight: note (small window)

**general** (20, m-, j-)
(09) TextViewer: from Open with...
(10) ArchiveViewer: file content
(11) KeyValuePage: long value
(12) FilemanagerBookInfo: book props (except description)
(13) InputText: clipboard (small window)
(14) screen_notification_menu_table: "Past notifications"
(15) FilemanagerBookInfo: list of books with hashdocsettings

**Comment by poire-z** [comment]
Looks sensible to me.
I'll probably group `(09) TextViewer: from Open with`... and `(10) ArchiveViewer: file content` in a **file content** category.

(No real life experience with bookmark details and notes to say how they are related - bookmark details being book content should be justified - dunno about the note as (probably smaller) user content if it should also be justified.)

Dunno about (14) and (15). (If any has some tabulation in it to indent/separate fields, I guess they should be in monospace, and more like code if they can be long.)

**Comment by hius07** [comment]
The latest commit adds clearing the TextViewer area when reinitting it after changing the monospace font.
It is necessary because of
https://github.com/koreader/koreader/blob/aabd6d7a2602e75d7e12d4d8e48f291f7e11275e/frontend/ui/widget/textboxwidget.lua#L126-L127
With non-monospace font the TextViewer window is taller than with monospace font (depending on the font size, for example by 3 pixels for font size 20 on my Kindle).

----------------------------

Another minor issue: when showing the TextViewer, we got a warning:
https://github.com/koreader/koreader/blob/aabd6d7a2602e75d7e12d4d8e48f291f7e11275e/frontend/ui/widget/button.lua#L494
for default buttons "top" and "bottom".
A warning occurs in init (before the TextViewer is drawn), and is caused with buttons refresh in
https://github.com/koreader/koreader/blob/aabd6d7a2602e75d7e12d4d8e48f291f7e11275e/frontend/ui/widget/textviewer.lua#L171-L181
Why to refresh undrawn buttons? Because of ScrollTextWidget init that contains
https://github.com/koreader/koreader/blob/aabd6d7a2602e75d7e12d4d8e48f291f7e11275e/frontend/ui/widget/scrolltextwidget.lua#L82
that contains
https://github.com/koreader/koreader/blob/aabd6d7a2602e75d7e12d4d8e48f291f7e11275e/frontend/ui/widget/scrolltextwidget.lua#L183
that calls buttons refresh.

`updateScrollBar()` was added to ScrollTextWidget init in https://github.com/koreader/koreader/pull/4084. I doubt if it is correct.

**Comment by poire-z** [comment]
> Another minor issue: when showing the TextViewer, we got a warning:

I don't get any warning (viewing Book description, or View HTML + CSS), with or without this PR...
Any specific context ?

I guess you could somehow wrap the :refresh in a `if self.dimen then` (or something like that, that exists only after the thing has been paintTo'ed).

About the first issue, I'm not sure about UIManager:show()'ing something that has been UIManager:close()'d, even if there's a :reinit(). @NiLuJe should give his thoughts (I think I met something like that recently, but i don't remember what it was and how it ended up being solved...)

**Comment by hius07** [comment]
> I don't get any warning

Debug mode?

**Comment by hius07** [comment]
> I'm not sure about UIManager:show()'ing something that has been UIManager:close()'d

It is not closed
https://github.com/koreader/koreader/blob/7368e461948c2b317f63b637e1b8c25256e44847/frontend/ui/widget/textviewer.lua#L548-L550

**Comment by hius07** [comment]
Please look at it
https://github.com/koreader/koreader/blob/aabd6d7a2602e75d7e12d4d8e48f291f7e11275e/frontend/ui/widget/scrolltextwidget.lua#L72-L82
I think `updateScrollBar()` should not be in init, initial scrollbar is good.
The scrollbar would be updated when jumping somewhere.

**Comment by poire-z** [comment]
> Debug mode?

Oh, right, indeed:
`12/10/23-18:24:59 DEBUG Button: table: 0x7f32b0a6e980 attempted a repaint in an unpainted frame!`

But if it's a DEBUG log - and not a WARN - it is fine :) even if the wording feels like it's the end of the world, I wouldn't worry - it's just @NiLuJe 's dark poetry style :)

> I think updateScrollBar() should not be in init, initial scrollbar is good.
The scrollbar would be updated when jumping somewhere.
EDIT: I removed it from init, works good in TextViewer / TextEditor.

It's been too long since I wrote it for me to justify it or suggest anything.
If we always open TextViewer with the cursor at top - and if the initial button state represents that, I guess it is ok to not call anything in :init() to update them.
But still, it calls some callback, you don't know if other users of ScrollTextWidget don't expect it to set their state - or force us to be sure the initial state is sane. So, something like just preventing the button:refresh() - the thing you are fighting - might be better than tweaking some upper more generic behaviour (that again, I can't really justify and swear it has some sense).

> > I'm not sure about UIManager:show()'ing something that has been UIManager:close()'d
> 
> It is not closed

I was talking about the popup, the UIManager:close() and :show() we see at the bottom of 7368e461948c.

**Comment by hius07** [comment]
> I was talking about the popup, the UIManager:close() and :show() we see at the bottom of [7368e46](https://github.com/koreader/koreader/commit/7368e461948c2b317f63b637e1b8c25256e44847).

Yes, I edited my reply.


> I close and reopen the ButtonDialog within itself, because it prevents refreshing the TextViewer. Works good.

I need to close it because otherwise old (tall) TextViewer is not cleared.

**Comment by NiLuJe** [comment]
Yep, as @poire-z said, that Button debug warning is mostly harmless ;).

File: frontend/ui/widget/textviewer.lua [hunk 3]
```diff
@@ -531,6 +542,93 @@ function TextViewer:handleTextSelection(text, hold_duration, start_idx, end_idx,
     end
 end
 
+function TextViewer:reinit()
+    local low, high = self.scroll_text_w.text_widget:getVisibleHeightRatios() -- try to keep position
+    local ratio = low == 0 and 0 or (low + high) / 2 -- if we are at the beginning, keep the first line visible
+    UIManager:setDirty(nil, "partial", self.frame.dimen)
```
Comment (NiLuJe) [inline frontend/ui/widget/textviewer.lua:548]: That's to clear the "old"/"current" instance in case the new one has different dimensions, right?

(No further comments, just wanted to confirm, as it's the exact right way to do that sort of thing ^^).

Comment (hius07) [reply frontend/ui/widget/textviewer.lua:548]: Yes, as I wrote above:
The latest commit adds clearing the TextViewer area when reinitting it after changing the monospace font.
With non-monospace font the TextViewer window is taller than with monospace font (depending on the font size, for example by 3 pixels for font size 20 on my Kindle).

**Comment by poire-z** [comment]
Another little thing I noticed earlier: when opening View HTML, we get a monospace font. In the popup menu, monospace is unchecked - and it needs 1 no-effect tap for it to rightly represent the state.

**Comment by hius07** [comment]
> In the popup menu, monospace is unchecked

This requires checking through all our monospace font aliases.
When we implement categories, we would not accept face/size from callers, and this would be fixed.

**Comment by poire-z** [comment]
> About the first issue, I'm not sure about UIManager:show()'ing something that has been UIManager:close()'d, even if there's a :reinit(). @NiLuJe should give his thoughts (I think I met something like that recently, but i don't remember what it was and how it ended up being solved...)

I remembered: https://github.com/koreader/koreader/pull/11126#issuecomment-1817950265
We still keep doing that in Trapper, but @NiLuJe felt the need to update the comment (still that dark poetry :):
https://github.com/koreader/koreader/blob/aabd6d7a2602e75d7e12d4d8e48f291f7e11275e/frontend/ui/trapper.lua#L181-L187

**Comment by poire-z** [comment]
> EDIT: Otherwise, currently we can pass the "monospace" flag from ViewHtml instead of font_face.

I think this should indeed be done that way. Either now or in the next PR.

We mostly set another font_face name when we indeed do want monospace..., so all the "infont", "smallinfofont" has become quite a bit useless... as it is always either NotoSans or DroidMonospace, and we use one of these keywords font when we're too lazy providing a size...
So, all our test could just work with 2 options: monospace=true/false, and a font size.

**Comment by hius07** [comment]
> May be a setDirty() to refresh that popup (which is over/on top of the TextViewer you reinit and redraw) could be enough? Unless it indeed need to go thru its :init() to update its state.

I close the popup not for refreshing itself (it is refreshed well via the new `checked_func`), but because opened popup prevents clearing old (tall) TextViewer window.

**Comment by poire-z** [comment]
> but because opened popup prevents clearing old (tall) TextViewer window.

"prevents clearing", you mean because of Lua stuff like references ?
Or just visually, because of incomplete refreshes regions?

I see that without the close/show, we get this border too thick, because it has remnants of the border of the previous taller TextViewer:
image

Is that what you've been fighting with this close/show ?
If yes, we usually fight this by combining the Geom/self.dimen of the old widget with the one of the new widget, getting a single larger Geom to give to setDirty.

**Comment by hius07** [comment]
Exactly, visual remnant of the taller TextViewer.
When the popup menu is opened,
https://github.com/koreader/koreader/blob/7368e461948c2b317f63b637e1b8c25256e44847/frontend/ui/widget/textviewer.lua#L548
doesn't work.

**Comment by poire-z** [comment]
It feels like you're in a situation similar (but probably simpler) to where we usually name the previous dimension `old_dimen`:
a complex one (you wouldn't need all that noise) for a :onShow(), which looks like what happens in your :reinit():
https://github.com/koreader/koreader/blob/aabd6d7a2602e75d7e12d4d8e48f291f7e11275e/frontend/ui/widget/configdialog.lua#L955-L974
and the simpler version:
https://github.com/koreader/koreader/blob/aabd6d7a2602e75d7e12d4d8e48f291f7e11275e/frontend/ui/widget/menu.lua#L1088-L1093

**Comment by poire-z** [comment]
> Is it okay that we explicitly specify the height of the widget but finally get the taller widget?

I guess so, if it's only small adjustments because of font size / line height.
(For the first display, we should compute everything so it fits in the provided height, and we add some padding if it's larger than a multiple integer of the line height - or reduce the provided height so it matches. Later, if we change font for one with a bit taller line height, I think it's better if the window grows a bit, than having to reduce the number of lines and add some larger paddings.)

**Comment by poire-z** [comment]
About the padding I mentionned: it must be only for the full screen TextEditor/InputDialog that I had to add some padding above and below the text widget, to ensure the fixed full screen height.
For some non-full screen widget, I guess it's best to avoid adding useless padding, and get variable height - and for the height to have the minimal change, it's best to just account for the small difference of line height, than having to reduce the number of lines of text and change the height by a full line-height.

**Comment by hius07** [comment]
Tried
```Lua
function TextViewer:reinit()
    local low, high = self.scroll_text_w.text_widget:getVisibleHeightRatios() -- try to keep position
    local ratio = low == 0 and 0 or (low + high) / 2 -- if we are at the beginning, keep the first line visible
    local old_dimen = self.frame.dimen:copy()
    self:init(true) -- do not add default buttons once more
    UIManager:setDirty(self, function()
        return "partial", old_dimen:combine(self.frame.dimen)
    end)
    self.scroll_text_w:scrollToRatio(ratio, ratio == 0)
end

```
Doesn't work with opened popup menu. The taller widget leftover remains on the screen.

**Comment by poire-z** [comment]
> Doesn't work with opened popup menu. The taller widget leftover remains on the screen.

Seems to work if you change:
```diff
-UIManager:setDirty(nil, function()
+UIManager:setDirty("all", function()
```

Looking at the code and comments of UIManager:setDirty(), it seems `nil` only works with "full" (ie. for small diagonal swipe).
Otherwise, it expects a widget - and it's usually the parent widget (`self.show_parent`) that we ask to repaint itself and its children (which should include the current widget) - and that works if the current widget is moved/updated inside the parent widget that usually doesn't change.
In your case of TextViewer, there is usually no reference to the parent passed when we create it (should it? probably painful to update it everywhere....). And I think in this specific case of Book description being possible called from anywhere including dispatcher, it may not be easy to always and everywhere pass the right parent.
So, using "all" (meaning asking the full stack of widget to repaint themselves) is probably ok (and not an issue for this specific rare case of toggling monospace.)
Pinging @NiLuJe for confirmation :)

**Comment by hius07** [comment]
Great, this works.

**Comment by hius07** [comment]
Again about
https://github.com/koreader/koreader/blob/aabd6d7a2602e75d7e12d4d8e48f291f7e11275e/frontend/ui/widget/scrolltextwidget.lua#L82
From the architecture point of view, is it correct that `widget:init()` contains `setDirty()` calls?

**Comment by poire-z** [comment]
> From the architecture point of view, is it correct that `widget:init()` contains `setDirty()` calls?

Probably not - and even more for "building block widget" like scrolltextwidget - only the upper real widget/window should setDirty itself, in :init() or probably better in :onShow().

Dunno if it could just not be called, or if it should so prev_low/high are set and the scroll_callback is called - haven't really looked at the other code using it if it would logically be expected or not - and if there is some needed consistency with `ScrollHtmlWidget:_updateScrollBar()` and `ScrollableContainer:_updateScrollBars()`.
May be it could accept a 2nd argument `initing=true` and it would skip all the refresh stuff if provided, and you would provide it only in :init().

**Comment by NiLuJe** [comment]
> Pinging @NiLuJe for confirmation :)

Please don't use `"all"` unless you *really* *really* *really* have to ;). It basically marks the *full* widget stack for *repainting* in addition to *refreshing*.

The trick that passing `nil` allows is to basically just request an arbitrary refresh *without* requesting any widgets to be repainted. You can use it with whichever region or refresh mode you want, it'll just enqueue a refresh request ;).

TL;DR: Either you have a top-level widget reference (e.g., something that was `UIManager:show`'n) and you can pass that to `setDirty`, or you don't, and pass a somewhat related reference if you have any for good conscience (but it won't do anything, and I murdered the debug guard that was warning about that because it was annoying), or you pass `nil` because you know the widget is being repainted by *something* anyway ;).

**Comment by NiLuJe** [comment]
With that said, as the first example you quoted in https://github.com/koreader/koreader/pull/11210#issuecomment-1849063651 demonstrate, when you go from a larger widget to a smaller one, you kinda need to repaint was was *behind* the old, larger widget, because if nothing actually repainted that region (and closing a widget doesn't technically paint anything), well, then that region of the screen buffer still contains the old widget ^^.

File: frontend/ui/widget/textviewer.lua [hunk 4]
```diff
@@ -69,26 +73,35 @@ local TextViewer = InputContainer:extend{
     find_centered_lines_count = 5, -- line with find results to be not far from the center
 }
 
-function TextViewer:init()
-    -- calculate window dimension
+local text_categories = {
```
Comment (poire-z) [inline frontend/ui/widget/textviewer.lua:76]: I use the word "category" when I introduced the idea, because it's the french word that came to my mind :) Dunno if it's really the best word for the setting and the variables. Sounds a bit odd reading that in English - for something that is more a kind or a type. @Frenzie @NiLuJe ? We will write and read:
```lua
    textviewer = TextViewer:new{
        title = _("Selection HTML"),
        text = html,
        text_category = "code", -- or "file_content", or "general"
```
Does this reads ok to you?

**Comment by hius07** [comment]
> Please don't use `"all"` unless you _really_ _really_ _really_ have to ;). It basically marks the _full_ widget stack for _repainting_ in addition to _refreshing_.

The problem here is that we have a top widget (popup menu) and need to refresh the widget behind it (TextViewer).

**Comment by poire-z** [comment]
> Please don't use "all" unless you really really really have to ;). 

TLDR: you have to, you can :)

**Comment by poire-z** [comment]
> The problem here is that we have a top widget (popup menu) and need to refresh the widget behind it (TextViewer).

I think that's not really it.
We have the TextViewer, not at top but as second to top (we can ignore the popup at top, it's not it that causes any issue).
You decrease the height of that TextViewer.
You need the thing below TextViewer (so, third to top :)) to repaint itself, so the little part of if shown in the few pixels that the new TextViewer does not cover get repainted.

And it's that thing below TextViewer that you would need to pass as a reference as show_parent, so only it get repainted.
That thing below TextViewer can be the fullscreen BookInformation widget, or another TextViewer (if viewing CSS from the ViewHTML), and I guess it can also be nothing, or seomthing else totally not accessible if the TextViewer is created from dispatcher/action/gestures...

**Comment by hius07** [comment]
> we can ignore the popup at top, it's not it that causes any issue

But closing the popup menu fixes the issue without "all".

**Comment by NiLuJe** [comment]
> But closing the popup menu fixes the issue without "all".

Because closing a widget marks every widget behind it as dirty, forcing a repaint ;).

The stuff quoted in https://github.com/koreader/koreader/pull/11210#issuecomment-1849063651 is tricky exactly because of that: it's the *same* widget instance that changes dimensions, so there's no close anywhere.

File: frontend/ui/widget/textviewer.lua [hunk 5]
```diff
@@ -531,6 +545,77 @@ function TextViewer:handleTextSelection(text, hold_duration, start_idx, end_idx,
...
 
+function TextViewer:reinit()
+    local text_settings = G_reader_settings:readSetting("textviewer_categories", {})
+    text_settings[self.text_category] =
+        { monospace_font = self.monospace_font, font_size = self.text_font_size, justified = self.justified }
+    local low, high = self.scroll_text_w.text_widget:getVisibleHeightRatios() -- try to keep position
+    local ratio = low == 0 and 0 or (low + high) / 2 -- if we are at the beginning, keep the first line visible
+    UIManager:setDirty("all", "partial", self.frame.dimen)
+    self:init(true) -- do not add default buttons once more
+    self:onShow()
```
[the remark this answers is not in the document]
Comment (hius07) [reply frontend/ui/widget/textviewer.lua]: Removed.

File: frontend/ui/widget/textviewer.lua [hunk 6]
```diff
@@ -69,26 +73,35 @@ local TextViewer = InputContainer:extend{
...
-        self.text_font_face = "x_smallinfofont"
-        if self.text_font_size == nil then
-            self.text_font_size = G_reader_settings:readSetting("dict_font_size")
-        end
+    self.width = self.width or screen_w - Screen:scaleBySize(30)
+    self.height = self.height or screen_h - Screen:scaleBySize(30)
+
+    if not reinit then
+        local categories = G_reader_settings:readSetting("textviewer_categories")
+        local text_settings = (categories and categories[self.text_category]) or text_categories[self.text_category]
```
[the remark this answers is not in the document]
Comment (hius07) [reply frontend/ui/widget/textviewer.lua]: Moved `text_categories` to self, for easy patching.

**Comment by hius07** [comment]
"Em space" is wider than the checkmark, causes truncation, will look at it tomorrow.

File: frontend/ui/widget/textviewer.lua (hunk 4 above)
Comment (poire-z) [reply frontend/ui/widget/textviewer.lua]: Re-pinging @Frenzie @NiLuJe

File: frontend/ui/widget/textviewer.lua [hunk 7]
```diff
@@ -67,28 +70,38 @@ local TextViewer = InputContainer:extend{
...
     default_hold_callback = nil, -- on each default button
     find_centered_lines_count = 5, -- line with find results to be not far from the center
+
+    text_category = "general",
+    text_categories = {
+        general      = { monospace_font = false, font_size = 20, justified = false },
+        file_content = { monospace_font = false, font_size = 20, justified = false },
+        book_info    = { monospace_font = false, font_size = 20, justified =  true },
+        bookmark     = { monospace_font = false, font_size = 20, justified =  true },
+        dictionary   = { monospace_font = false, font_size = 20, justified = false },
```
Comment (poire-z) [inline frontend/ui/widget/textviewer.lua:80]: This "dictionary" does not impact any dictionary result I think (we don't use TextViewer for them). It would be used only for translator currently.
So, may be "lookup_result" or "lookup" ?

File: frontend/ui/widget/textviewer.lua (hunk 4 above)
Comment (Frenzie) [reply frontend/ui/widget/textviewer.lua]: I think type would be the most obvious and variety would probably read a little more natural than category but I'm fine with category too.

Comment (NiLuJe) [reply frontend/ui/widget/textviewer.lua]: Yup, my first thought would have been `type` too, but this is fine ;).

**Comment by hius07** [comment]
BookInfo module will get text types in the forthcoming PR introducing "set screenshot as book cover".
The rest of the modules use default "general" text type (and will be tuned after using in the field if need be).

**Review by poire-z — APPROVED** [review]

**Comment by hius07** [comment]
Current text types:

**code** (16, m+, j-)
(01) ViewHtml: code
(02) ReaderTypography: "Language tags (and hyphenation dictionaries) used since start up"
(03) PatchManagement: code

**lookup** (20, m-, j-)
(04) Translator: translation

**book_info** (20, m-, j+)
(05) FilemanagerBookInfo: book description
(06) OPDSBrowser: book information

**bookmark** (20, m-, j+)
(07) ReaderBookmark: bookmark details

**file_content** (20, m-, j-)
(09) TextViewer: from Open with...
(10) ArchiveViewer: file content

**general** (default) (20, m-, j-)
(08) ReaderHighlight: note (small window, no menu)
(11) KeyValuePage: long value
(12) FilemanagerBookInfo: book props (except description)
(13) InputText: clipboard (small window, no menu)
(14) screen_notification_menu_table: "Past notifications"
(15) FilemanagerBookInfo: list of books with hashdocsettings

**Comment by poire-z** [comment]
Just noticed that if we launch Book description from the long-press on a file in FM, we get the text justified - but if we launch it from the Book information KeyValuePage when in Reader, we get the text non-justified.

**Comment by hius07** [comment]
> if we launch it from the Book information KeyValuePage when in Reader, we get the text non-justified

It is shown as long KVP value, text_type = "general".
Must be fixed by adding `entry.callback` to the Description kvp item.
